wurst/file/Base64.wurst

```diff
@@ -0,0 +1,270 @@
+    > let byte = bytes.readByte()
+    > let number = bytes.readInt()
+
+    This package also provides convenient methods for ChunkedStrings and ByteBuffers.
+
+    Encode Data:
+    > let data = bytes.encodeBase64()
+
+    Decode Data:
+    > let bytes = data.decodeBase64()
+*/
+
+/**
+    Specifies how many characters to encode per a single `execute()` call.
+    This value has been tuned to work under all optimization settings and
+    with stacktraces included.
+**/
+@configurable public constant ENCODES_PER_ROUND = 1000
+/**
+    Specifies how many chunks to decode per a single `execute()` call.
+    This value has been tuned to work under all optimization settings and
+    with stacktraces included.
+**/
+@configurable public constant DECODES_PER_ROUND = 100
+
+// RFC 4648 compliant Base64 charmap.
+constant CHARMAP = [
+    "A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P",
+    "Q","R","S","T","U","V","W","X","Y","Z","a","b","c","d","e","f",
+    "g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v",
+    "w","x","y","z","0","1","2","3","4","5","6","7","8","9","+","/"
+]
+
+// RFC 4648 compliant Base64 reverse charmap.
+constant REVERSE_CHARMAP = [
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 62, 0, 0, 0, 63, 52, 53, 54,
+    55, 56, 57, 58, 59, 60, 61, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2,
+    3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
+    20, 21, 22, 23, 24, 25, 0, 0, 0, 0, 0, 0, 26, 27, 28, 29, 30,
+    31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
+    48, 49, 50, 51, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+]
+
+public class Base64Encoder
+    private ChunkedString chunkedString
+
+    // We append only one char each time, and strings in WC3 get fully copied on each concatenation.
+    // To optimize it and lower the times of copying quite big chunks, append by smaller chunks.
+    private var stringBuffer = ""
+    private var chars = 0
+    private static constant MAX_CHARS = 32
+
+    private var buffer = 0
+    private var bytes = 0
+    private static constant MAX_BYTES = 3
+
+    construct()
+        this.chunkedString = new ChunkedString()
+
+    construct(int maxChunkLength)
+        this.chunkedString = new ChunkedString(maxChunkLength)
+
+    ondestroy
+        if chunkedString != null // The string is null after calling `intoData()`.
+            destroy chunkedString
+
+    // It's way too much overhead in dispatching two lambdas per each byte, 'inlining' them to deal with this.
```

Comment:
> It's way too much overhead

Justification?

> 'inlining' them

Who is inlining them, where?

Reply:
You just have to know what does it compile to. A big switch with elseifs per each possible instance and a TriggerExecute. It's obvious that we do not want to do 10 ifs and a TriggerExecute call per each byte.
That was meant to describe why I'm not using stdlib's executeWhile here and reimplement the functionality, by 'inlining' the lambdas into the executeWhile. I'll make the description better.

Reply:
IMO you can just delete this comment